Introduce max_total_package_size that accounts for total size

akmysti opened the pull request:
## Description 

Introduce `max_total_linkage_size`: a protocol-gated cap on the total
on-disk footprint of a Move package plus all of its transitive
dependencies, enforced at publish and upgrade time.

Previously we had two independent caps — `max_move_package_size` (per
package) and `max_package_dependencies` (max transitive deps) — and the
effective upper bound on a loaded linkage was their product. That made
it hard to relax either one without blowing up the worst-case linkage
size, which is what actually matters for storage, loading, and the
verifier.

This PR adds a single total-linkage cap so the two per-unit limits can
be raised independently of the aggregate bound. At v123:

- `max_move_package_size`: 100 KB → 1 MB
- `max_package_dependencies`: 32 → 100
- `max_total_linkage_size`: new, set to `32 * 100 * 1024` (the product
  of the *previous* limits), so the worst-case total footprint does
  not regress relative to v122.

## Protocol gating

All three changes ship under a new **protocol version 123**:

- `MAX_PROTOCOL_VERSION` bumped 122 → 123.
- New `123 => { … }` arm in `ProtocolConfig::get_for_version_impl`.
- `max_total_linkage_size_as_option()` is used at the call site so the
  check is a no-op on pre-v123 configs — v122 and earlier are
  byte-identical to today.

Snapshots regenerated:

- `version_122.snap` unchanged — still `max_move_package_size: 102400`,
  `max_package_dependencies: 32`, no `max_total_linkage_size`.
- `version_123.snap` (and Mainnet/Testnet variants) show
  `max_move_package_size: 1048576`, `max_package_dependencies: 100`,
  `max_total_linkage_size: 3276800`.

## Implementation

- `MovePackage::from_module_iter_with_type_origin_table` (the single
  site shared by `new_initial` and `new_upgraded`) now sums the size
  of the new package plus each transitive dep and, when
  `max_total_linkage_size_as_option()` is `Some`, rejects publishes /
  upgrades whose total exceeds the cap with `MovePackageTooBig`.
- `build_linkage_table` returns `(linkage_table, total_deps_size)` so
  the caller gets the dep footprint without a second pass.
- `MovePackage::new` is unchanged in signature and still enforces the
  per-package `max_move_package_size`; the total-linkage check lives
  one level up, where the transitive deps are in scope.
- Field doc on `max_total_linkage_size` names the real enforcement
  site instead of a `TODO`.

## Tests

New unit tests in `sui-types/src/move_package.rs` — all go end-to-end
through `MovePackage::new_initial` against a real `ProtocolConfig` at
v123, with per-field `set_*_for_testing` / `disable_*_for_testing`
overrides to keep the synthetic modules small:

- `test_max_total_linkage_size_check_passes` — package + dep well
  under the cap; creation succeeds.
- `test_max_total_linkage_size_check_fails` — cap sized just below
  `dep.size() + pkg.size()`; returns `MovePackageTooBig` with the
  aggregate size and the configured max.
- `test_max_total_linkage_size_disabled_allows_oversize` — when the
  config field is `None`, the check is skipped entirely (guards the
  pre-v123 gate path).

## Test plan

- [x] `cargo nextest run -p sui-types --lib move_package::` — 3/3 new
      tests pass; no regressions in existing `move_package` tests.
- [x] `cargo nextest run -p sui-protocol-config` — 7/7 pass; v123
      snapshots accepted, v122 and earlier unchanged.
- [x] `cargo clippy -p sui-types -p sui-protocol-config --tests
      --no-deps` — clean.
- [ ] `/protocol-config` review to double-check v123 gating before
      merge.
- [ ] Full `cargo simtest -p sui-e2e-tests` before merge.

## Follow-ups (not in this PR)

- Introduce a dedicated `MovePackageLinkageTooBig` variant of
  `ExecutionErrorKind` so operators/tools can distinguish "this one
  package is too big" from "the transitive linkage is too big."
  Deferred because it changes the on-wire execution status enum and
  requires proto / SDK conversion + YAML snapshot updates.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] gRPC:
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

cgswords commented on crates/sui-protocol-config/src/lib.rs:
```diff
@@ -3906,7 +3910,11 @@ impl ProtocolConfig {
                 }
-                91 => {}
+                91 => {
+                    cfg.max_package_dependencies = Some(100); // Relaxed
+                    cfg.max_move_package_size = Some(1024 * 1024); // Relaxed to 1 MB
+                    cfg.max_total_package_size = Some(32 * 100 * 1024); // Product of the two from before.
```
```suggestion
                    cfg.max_package_dependencies = Some(100);
                    cfg.max_move_package_size = Some(1024 * 1024);
                    cfg.max_total_package_size = Some(32 * 100 * 1024); // Product of previous package and dependency limits.
```

cgswords commented on crates/sui-protocol-config/src/lib.rs:
```diff
@@ -996,6 +996,9 @@ pub struct ProtocolConfig {
     /// Maximum size of a Move package object, in bytes. Enforced by the Sui adapter at the end of a publish transaction.
     max_move_package_size: Option<u64>,
 
+    // Goal is to replace max_move_package_size and max_total_package_size with this.
+    max_total_package_size: Option<u64>,
```
All limits need a doc comment.
```suggestion
    /// Maximum size of a Move package with its dependencies, in bytes. This limits the overall size of a loaded package and its dependencies.
    // Enforced by the <WHO?> at <WHERE?>.
    max_total_linkage_size: Option<u64>,
```

cgswords reviewed (changes requested):
This does not actually enforce the new max size, so we cannot land it (along with the increases) yet.

akmysti commented on crates/sui-protocol-config/src/lib.rs:
```diff
@@ -996,6 +996,10 @@ pub struct ProtocolConfig {
     max_move_package_size: Option<u64>,
 
+    /// Maximum size of a Move package with its dependencies, in bytes. This limits the overall size of a loaded package and its dependencies.
+    // Enforced by the <WHO?> at <WHERE?>.
```
TODO

akmysti commented on crates/sui-protocol-config/src/lib.rs:
```diff
@@ -3906,7 +3911,11 @@ impl ProtocolConfig {
+                91 => {
+                    cfg.max_package_dependencies = Some(100);
+                    cfg.max_move_package_size = Some(1024 * 1024);
+                    cfg.max_total_package_size = Some(32 * 100 * 1024); // Product of previous package and dependency limits.
```
TODO: FInd all the places where the other two are enforced and use the new cfg instead.

akmysti committed:
Introduce max_total_package_size that accounts for total size

Rather than having two static metric that can be relaxed much more without blowing
up the total size which is what matters eventually

akmysti committed:
Protocol-gate max_total_linkage_size at v123

The prior change on this branch placed the new total-linkage-size cap
(and bumps to max_move_package_size / max_package_dependencies) under
protocol version 98, which is already live. Mutating a shipped version
is consensus-breaking, so move all three changes to a new v123 arm and
bump MAX_PROTOCOL_VERSION accordingly.

While here, tighten the implementation:

- Drop the public `new_with_linkage_size` helper (its two correlated
  Option params were a footgun) and restore `MovePackage::new` to its
  single-cap form. The total-linkage check now lives inline in
  `from_module_iter_with_type_origin_table`, the single site shared by
  `new_initial` and `new_upgraded`, gated by
  `max_total_linkage_size_as_option()` so the check is a no-op on
  pre-v123 configs.
- Replace the `TODO: Enforced by the <WHO?> at <WHERE?>.` placeholder
  on the field doc with the real enforcement pointer.
- Rewrite the unit tests to go end-to-end through `new_initial` with a
  real `ProtocolConfig` at v123 (using `set_*_for_testing` to size the
  caps to the small test modules). Add a "disabled" test that confirms
  the check is a no-op when `max_total_linkage_size` is None.

Test plan:
- cargo nextest run -p sui-types --lib move_package::
  3/3 pass (passes / fails / disabled).
- cargo nextest run -p sui-protocol-config
  7/7 pass; regenerated v123 snapshots (Unknown / Mainnet / Testnet)
  show 100 / 1048576 / 3276800; v122 snapshot unchanged at 32 / 102400
  with no max_total_linkage_size field.
- cargo clippy -p sui-types -p sui-protocol-config --tests --no-deps
  clean.

Follow-ups (not in this commit):
- Consider a dedicated `MovePackageLinkageTooBig` ExecutionErrorKind so
  operators can tell this from the per-package size cap. Deferred
  because it changes the on-wire execution status enum (proto + SDK
  conversions + YAML snapshots).
- Run /protocol-config to validate the v123 gating before opening the
  PR.

akmysti committed:
Add more failing tests

test_max_total_linkage_size_fails_on_many_small_deps
test_max_total_linkage_size_fails_on_single_huge_dep
test_max_total_linkage_size_enforced_on_upgrade

akmysti committed:
rustfmt

akmysti committed:
Update snapshots for protocol version 123

simtest CI surfaced two snapshot tests that pin the max protocol
version:

- sui-swarm-config populated_genesis_snapshot_matches-2: now records
  protocol_version: 123.
- sui-core execution_time_estimator snapshot_tests: adds the v123
  observer snapshot alongside the existing per-version snaps.

WayneAl commented:
When will this config update on mainnet?
